File: digest_system/runtime/reporting.py
"""Run reporting: the per-run cost summary and the readback of measured stage usage.

Python port of ``src/runtime/reporting.mjs``.
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Mapping, Sequence

from ..editorial.stages import STAGES_V2
from .artifacts import ROOT, RUNS_DIRECTORY, read_json, stage_directory
from .costs import PRICING, billing_band, cost_for_band


def build_run_summary(
    *,
    run_id: str,
    digest_id: str,
    style: str,
    corpus_policy: str | None,
    stages: Sequence[Mapping[str, Any]],
    extra: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Build the per-run cost record from the measured usage of each stage, and derive the
    counterfactuals that tell us whether scheduling and caching are pulling their weight."""
    hit = miss = output = reasoning = 0
    seconds = 0.0
    actual = all_off_peak = all_peak = 0.0
    per_stage: list[dict[str, Any]] = []

    for stage in stages:
        hit += stage["hit"]
        miss += stage["miss"]
        output += stage["output"]

        # Price each model call in its own billing band. A stage that made a second attempt can
        # straddle a peak boundary, so banding the aggregate by the stage's start would misprice
        # it; the per-attempt measurement is preferred whenever the caller supplies one.
        priced_calls = [
            attempt
            for attempt in (stage.get("attempts") or [])
            if attempt.get("completed") and attempt.get("started_at")
        ]
        calls = (
            [
                {
                    "startedAt": attempt["started_at"],
                    "seconds": attempt.get("seconds", 0),
                    "hit": attempt.get("cache_hit_tokens", 0),
                    "miss": attempt.get("cache_miss_tokens", 0),
                    "output": attempt.get("output_tokens", 0),
                    "reasoning": attempt.get("reasoning_tokens", 0),
                }
                for attempt in priced_calls
            ]
            if priced_calls
            else [
                {
                    "startedAt": stage["startedAt"],
                    "seconds": stage["seconds"],
                    "hit": stage["hit"],
                    "miss": stage["miss"],
                    "output": stage["output"],
                    "reasoning": stage["reasoning"],
                }
            ]
        )

        stage_cost = stage_off_peak = stage_peak = 0.0
        for call in calls:
            band = billing_band(call["startedAt"])
            stage_cost += cost_for_band(band, call)
            stage_off_peak += cost_for_band("off-peak", call)
            stage_peak += cost_for_band("peak", call)
        actual += stage_cost
        all_off_peak += stage_off_peak
        all_peak += stage_peak
        reasoning += stage["reasoning"]
        seconds += stage["seconds"]

        bands = sorted({billing_band(call["startedAt"]) for call in calls})

        entry: dict[str, Any] = {
            "stage": stage["name"],
            "started_at": stage["startedAt"],
            "completed_at": stage["completedAt"],
            "seconds": round(stage["seconds"], 2),
        }
        if isinstance(stage.get("model_seconds"), (int, float)):
            entry["model_seconds"] = round(stage["model_seconds"], 2)
        if isinstance(stage.get("attempt_count"), int):
            entry["attempt_count"] = stage["attempt_count"]
        if len(calls) > 1:
            entry["attempts"] = [
                {
                    "attempt": index + 1,
                    "started_at": call["startedAt"],
                    "seconds": round(call["seconds"], 2),
                    "billing_band": billing_band(call["startedAt"]),
                    "cache_hit_tokens": call["hit"],
                    "cache_miss_tokens": call["miss"],
                    "output_tokens": call["output"],
                    "reasoning_tokens": call["reasoning"],
                    "cost_usd": round(cost_for_band(billing_band(call["startedAt"]), call), 6),
                }
                for index, call in enumerate(calls)
            ]
        entry.update(
            {
                # `mixed` for a stage whose attempts landed in different bands, so a run that
                # straddles a boundary is visible rather than silently averaged.
                "billing_band": bands[0] if len(bands) == 1 else "mixed",
                "cache_hit_tokens": stage["hit"],
                "cache_miss_tokens": stage["miss"],
                "output_tokens": stage["output"],
                "reasoning_tokens": stage["reasoning"],
                "cache_hit_ratio": (
                    round(stage["hit"] / (stage["hit"] + stage["miss"]), 4)
                    if stage["hit"] + stage["miss"]
                    else None
                ),
                "cost_usd": round(stage_cost, 6),
            }
        )
        if stage.get("provenance"):
            entry["provenance"] = stage["provenance"]
        per_stage.append(entry)

    billed_bands = sorted({stage["billing_band"] for stage in per_stage})

    summary: dict[str, Any] = {
        "schema_version": 2,
        "run_id": run_id,
        "digest_id": digest_id,
        "style": style,
        "corpus_policy": corpus_policy,
        "started_at": per_stage[0]["started_at"] if per_stage else None,
        "completed_at": per_stage[-1]["completed_at"] if per_stage else None,
        "billing_band": billed_bands[0] if len(billed_bands) == 1 else "mixed",
        "total_seconds": round(seconds, 2),
        "tokens": {
            "cache_hit": hit,
            "cache_miss": miss,
            "output": output,
            "reasoning": reasoning,
            "total_input": hit + miss,
            "total": hit + miss + output,
        },
        "cost_usd": {
            "actual": round(actual, 6),
            "if_all_off_peak": round(all_off_peak, 6),
            "if_all_peak": round(all_peak, 6),
            "if_nothing_cached": round(
                ((hit + miss) / 1e6) * PRICING["cacheMiss"]["offPeak"] + (output / 1e6) * PRICING["output"]["offPeak"],
                6,
            ),
        },
        "stages": per_stage,
    }
    summary.update(extra or {})
    return summary


def format_cost_summary(summary: Mapping[str, Any]) -> str:
    cost = summary["cost_usd"]
    tokens = summary["tokens"]
    return (
        f"run cost: ${cost['actual']:.4f} ({summary['billing_band']}, {summary['total_seconds']:.0f}s) | "
        f"tokens {tokens['total']:,} = {tokens['cache_hit']:,} hit + {tokens['cache_miss']:,} miss + {tokens['output']:,} out | "
        f"off-peak would be ${cost['if_all_off_peak']:.4f}, all-peak ${cost['if_all_peak']:.4f}, no-cache ${cost['if_nothing_cached']:.4f}"
    )


def read_stage_attempts(run_id: str, stage: Any, *, root: Path | None = None) -> list[dict[str, Any]]:
    """Read one stage's measured usage, aggregating **every** attempt it made.

    A stage can make more than one model call, because a validation failure earns one
    correction attempt. Reading only ``attempt-1`` would report the cost and duration of the
    rejected call and silently omit the call that produced the artifact.
    """
    base = root or ROOT
    attempts_dir = stage_directory(run_id, stage.name, base) / "attempts"
    if not attempts_dir.is_dir():
        return []
    numbers = sorted(
        int(match.group(1))
        for entry in attempts_dir.iterdir()
        if entry.is_dir() and (match := re.match(r"^attempt-(\d+)$", entry.name))
    )

    attempts = []
    for number in numbers:
        attempt_dir = attempts_dir / f"attempt-{number}"
        try:
            attempt = read_json(attempt_dir / "attempt.json")
            completed = read_json(attempt_dir / "completed.json")
        except (OSError, ValueError):
            # An attempt that did not complete contributes no measurement, but it is still
            # recorded as an attempt so a reviewer can see that a call was made and failed.
            attempts.append({"attempt": number, "completed": False})
            continue
        usage = completed.get("usage") or {}
        attempts.append(
            {
                "attempt": number,
                "completed": True,
                "started_at": attempt.get("started_at"),
                "completed_at": completed.get("completed_at"),
                "seconds": _seconds_between(attempt.get("started_at"), completed.get("completed_at")),
                "cache_hit_tokens": completed.get("cache_hit_tokens", 0),
                "cache_miss_tokens": completed.get("cache_miss_tokens", 0),
                "output_tokens": usage.get("completion_tokens", 0),
                "reasoning_tokens": (usage.get("completion_tokens_details") or {}).get("reasoning_tokens", 0),
                "finish_reason": completed.get("finish_reason"),
                "validation_correction": bool(attempt.get("validation_correction")),
            }
        )
    return attempts


def _seconds_between(start: Any, end: Any) -> float:
    from datetime import datetime

    def parse(value: Any) -> datetime:
        text = str(value)
        if text.endswith("Z"):
            text = text[:-1] + "+00:00"
        return datetime.fromisoformat(text)

    return (parse(end) - parse(start)).total_seconds()


def read_measured_stages_v2(run_id: str, *, root: Path | None = None) -> list[dict[str, Any]]:
    measured = []
    for stage in STAGES_V2:
        attempts = read_stage_attempts(run_id, stage, root=root)
        completed = [attempt for attempt in attempts if attempt.get("completed")]
        if not completed:
            continue
        first, last = completed[0], completed[-1]

        def total(key: str) -> Any:
            return sum(attempt.get(key, 0) or 0 for attempt in completed)

        measured.append(
            {
                "name": stage.name,
                "startedAt": first["started_at"],
                "completedAt": last["completed_at"],
                # The stage's wall time, from its first attempt starting to its last one
                # finishing. This includes the validation between attempts.
                "seconds": _seconds_between(first["started_at"], last["completed_at"]),
                "model_seconds": total("seconds"),
                "attempt_count": len(completed),
                "attempts": attempts,
                "hit": total("cache_hit_tokens"),
                "miss": total("cache_miss_tokens"),
                "output": total("output_tokens"),
                "reasoning": total("reasoning_tokens"),
                "provenance": "python-adapter" if stage.executor == "evaluation" else "runner",
            }
        )
    return measured


def read_stage_records_v2(run_id: str, *, root: Path | None = None) -> dict[str, Any] | None:
    base = root or ROOT
    try:
        return read_json(base / RUNS_DIRECTORY / run_id / "stage-records.json")
    except (OSError, ValueError):
        return None


__all__ = [
    "build_run_summary",
    "format_cost_summary",
    "read_stage_attempts",
    "read_measured_stages_v2",
    "read_stage_records_v2",
]

File: src/runtime/reporting.mjs
// Run reporting: the per-run cost summary and the readback of measured stage usage.

import { readdir } from "node:fs/promises";
import path from "node:path";

import { RUNS_DIRECTORY, ROOT, readJson, stageDirectory } from "./artifacts.mjs";
import { billingBand, costForBand, PRICING } from "./costs.mjs";
import { STAGES_V2 } from "../editorial/stages.mjs";

// Build the per-run cost record from the measured usage of each stage, and derive the
// counterfactuals that tell us whether scheduling and caching are pulling their weight.
export function buildRunSummary({ runId, digestId, style, corpusPolicy, stages, extra = {} }) {
  let hit = 0, miss = 0, output = 0, reasoning = 0, seconds = 0;
  let actual = 0, allOffPeak = 0, allPeak = 0;
  const perStage = [];

  for (const s of stages) {
    hit += s.hit; miss += s.miss; output += s.output;

    // Price each model call in its own billing band. A stage that made a second attempt can
    // straddle a peak boundary, so banding the aggregate by the stage's start would misprice it;
    // the per-attempt measurement is preferred whenever the caller supplies one, and the stage
    // aggregate is the fallback for a single-attempt stage or a historical run.
    const pricedCalls = (Array.isArray(s.attempts) ? s.attempts : []).filter(
      (attempt) => attempt.completed && attempt.started_at,
    );
    const calls = pricedCalls.length
      ? pricedCalls.map((attempt) => ({
          startedAt: attempt.started_at,
          seconds: attempt.seconds ?? 0,
          hit: attempt.cache_hit_tokens ?? 0,
          miss: attempt.cache_miss_tokens ?? 0,
          output: attempt.output_tokens ?? 0,
          reasoning: attempt.reasoning_tokens ?? 0,
        }))
      : [{ startedAt: s.startedAt, seconds: s.seconds, hit: s.hit, miss: s.miss, output: s.output, reasoning: s.reasoning }];

    let stageCost = 0;
    let stageOffPeak = 0;
    let stagePeak = 0;
    for (const call of calls) {
      const band = billingBand(call.startedAt);
      stageCost += costForBand(band, call);
      stageOffPeak += costForBand("off-peak", call);
      stagePeak += costForBand("peak", call);
    }
    actual += stageCost;
    allOffPeak += stageOffPeak;
    allPeak += stagePeak;
    reasoning += s.reasoning;
    seconds += s.seconds;

    const bands = [...new Set(calls.map((call) => billingBand(call.startedAt)))];

    perStage.push({
      stage: s.name,
      started_at: s.startedAt,
      completed_at: s.completedAt,
      seconds: Number(s.seconds.toFixed(2)),
      // The sum of this stage's model calls, when the caller reports it separately. The stage's
      // wall time above includes any validation between attempts, so the two differ by the cost
      // of the checks — which is what a reviewer wants to see when a stage retries.
      ...(typeof s.model_seconds === "number" ? { model_seconds: Number(s.model_seconds.toFixed(2)) } : {}),
      ...(typeof s.attempt_count === "number" ? { attempt_count: s.attempt_count } : {}),
      ...(calls.length > 1 ? { attempts: calls.map((call, index) => ({
        attempt: index + 1,
        started_at: call.startedAt,
        seconds: Number(call.seconds.toFixed(2)),
        billing_band: billingBand(call.startedAt),
        cache_hit_tokens: call.hit,
        cache_miss_tokens: call.miss,
        output_tokens: call.output,
        reasoning_tokens: call.reasoning,
        cost_usd: Number(costForBand(billingBand(call.startedAt), call).toFixed(6)),
      })) } : {}),
      // `mixed` for a stage whose attempts landed in different bands, so a run that straddles a
      // boundary is visible rather than silently averaged.
      billing_band: bands.length === 1 ? bands[0] : "mixed",
      cache_hit_tokens: s.hit,
      cache_miss_tokens: s.miss,
      output_tokens: s.output,
      reasoning_tokens: s.reasoning,
      cache_hit_ratio: s.hit + s.miss ? Number((s.hit / (s.hit + s.miss)).toFixed(4)) : null,
      cost_usd: Number(stageCost.toFixed(6)),
      ...(s.provenance ? { provenance: s.provenance } : {}),
    });
  }

  const billedBands = [...new Set(perStage.map((s) => s.billing_band))];

  return {
    schema_version: 2,
    run_id: runId,
    digest_id: digestId,
    style,
    corpus_policy: corpusPolicy ?? null,
    started_at: perStage[0]?.started_at ?? null,
    completed_at: perStage[perStage.length - 1]?.completed_at ?? null,
    // Recorded explicitly so a later analysis never has to re-derive the band, and so a
    // run that straddles a boundary is visible rather than silently averaged.
    billing_band: billedBands.length === 1 ? billedBands[0] : "mixed",
    total_seconds: Number(seconds.toFixed(2)),
    tokens: {
      cache_hit: hit,
      cache_miss: miss,
      output,
      reasoning,
      total_input: hit + miss,
      total: hit + miss + output,
    },
    cost_usd: {
      actual: Number(actual.toFixed(6)),
      // Same work, priced entirely in the cheaper or the dearer band.
      if_all_off_peak: Number(allOffPeak.toFixed(6)),
      if_all_peak: Number(allPeak.toFixed(6)),
      // Same work with no cache reuse at all, priced off-peak.
      if_nothing_cached: Number((((hit + miss) / 1e6) * PRICING.cacheMiss.offPeak + (output / 1e6) * PRICING.output.offPeak).toFixed(6)),
    },
    stages: perStage,
    ...extra,
  };
}

export function formatCostSummary(summary) {
  const c = summary.cost_usd;
  const t = summary.tokens;
  return `run cost: $${c.actual.toFixed(4)} (${summary.billing_band}, ${summary.total_seconds.toFixed(0)}s) | ` +
    `tokens ${t.total.toLocaleString()} = ${t.cache_hit.toLocaleString()} hit + ${t.cache_miss.toLocaleString()} miss + ${t.output.toLocaleString()} out | ` +
    `off-peak would be $${c.if_all_off_peak.toFixed(4)}, all-peak $${c.if_all_peak.toFixed(4)}, no-cache $${c.if_nothing_cached.toFixed(4)}`;
}

/**
 * Read one stage's measured usage, aggregating **every** attempt it made.
 *
 * A stage can now make more than one model call, because a validation failure earns one
 * correction attempt. Reading only `attempt-1` would therefore report the cost and duration of
 * the rejected call and silently omit the call that produced the artifact — so the run summary,
 * the cost ledger and every later cost comparison would understate exactly the stages whose
 * instructions are being tuned. Each attempt's own measurement is preserved alongside the
 * aggregate, and the stage's wall time is kept distinct from the sum of its model calls, since
 * the difference is the time spent validating between them.
 */
async function readStageAttempts(runId, stage) {
  const attemptsDir = path.join(stageDirectory(runId, stage.name), "attempts");
  const entries = await readdir(attemptsDir, { withFileTypes: true }).catch(() => []);
  const numbers = entries
    .filter((entry) => entry.isDirectory())
    .map((entry) => /^attempt-(\d+)$/.exec(entry.name))
    .filter(Boolean)
    .map((match) => Number(match[1]))
    .sort((a, b) => a - b);

  const attempts = [];
  for (const number of numbers) {
    const attemptDir = path.join(attemptsDir, `attempt-${number}`);
    let attempt;
    let completed;
    try {
      attempt = await readJson(path.join(attemptDir, "attempt.json"));
      completed = await readJson(path.join(attemptDir, "completed.json"));
    } catch {
      // An attempt that did not complete contributes no measurement, but it is still recorded
      // as an attempt so a reviewer can see that a call was made and failed.
      attempts.push({ attempt: number, completed: false });
      continue;
    }
    const usage = completed.usage ?? {};
    attempts.push({
      attempt: number,
      completed: true,
      started_at: attempt.started_at,
      completed_at: completed.completed_at,
      seconds: (new Date(completed.completed_at) - new Date(attempt.started_at)) / 1000,
      cache_hit_tokens: completed.cache_hit_tokens ?? 0,
      cache_miss_tokens: completed.cache_miss_tokens ?? 0,
      output_tokens: usage.completion_tokens ?? 0,
      reasoning_tokens: usage.completion_tokens_details?.reasoning_tokens ?? 0,
      finish_reason: completed.finish_reason ?? null,
      validation_correction: Boolean(attempt.validation_correction),
    });
  }
  return attempts;
}

export async function readMeasuredStagesV2(runId) {
  const measured = [];
  for (const stage of STAGES_V2) {
    const attempts = await readStageAttempts(runId, stage);
    const completed = attempts.filter((attempt) => attempt.completed);
    if (completed.length === 0) continue;

    const first = completed[0];
    const last = completed[completed.length - 1];
    const sum = (key) => completed.reduce((total, attempt) => total + (attempt[key] ?? 0), 0);
    const modelSeconds = sum("seconds");

    measured.push({
      name: stage.name,
      startedAt: first.started_at,
      completedAt: last.completed_at,
      // The stage's wall time, from its first attempt starting to its last one finishing. This
      // includes the validation between attempts, so it is not the same as `model_seconds`.
      seconds: (new Date(last.completed_at) - new Date(first.started_at)) / 1000,
      model_seconds: modelSeconds,
      attempt_count: completed.length,
      attempts,
      hit: sum("cache_hit_tokens"),
      miss: sum("cache_miss_tokens"),
      output: sum("output_tokens"),
      reasoning: sum("reasoning_tokens"),
      provenance: stage.executor === "evaluation" ? "python-adapter" : "runner",
    });
  }
  return measured;
}

export async function readStageRecordsV2(runId) {
  try {
    return await readJson(path.join(ROOT, RUNS_DIRECTORY, runId, "stage-records.json"));
  } catch {
    return null;
  }
}